src/Common/src/Common/Configuration/StringExtensions.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.Configuration;
+
+namespace Steeltoe.Common.Configuration;
+
+public static class StringExtensions
+{
+    private const string DotDelimiterString = ".";
+    private const char DotDelimiterChar = '.';
+    private const char UnderscoreDelimiterChar = '_';
+    private const char EscapeChar = '\\';
+    private const string EscapeString = "\\";
+
+    private static readonly Regex ArrayRegex = new(@"(\[[0-9]+\])*$");
+
+    public static string AsDotNetConfigurationKey(this string key)
+    {
+        if (string.IsNullOrEmpty(key))
+        {
+            return key;
+        }
+
+        IEnumerable<string> split = UniversalHierarchySplit(key);
+        var sb = new StringBuilder();
+
+        foreach (string keyPart in split.Select(ConvertArrayKey))
+        {
+            sb.Append(keyPart);
+            sb.Append(ConfigurationPath.KeyDelimiter);
+        }
+
+        return sb.ToString(0, sb.Length - 1);
+    }
+
+    private static IEnumerable<string> UniversalHierarchySplit(string source)
+    {
+        ArgumentGuard.NotNull(source);
+
+        var result = new List<string>();
+
+        int segmentStart = 0;
+
+        for (int i = 0; i < source.Length; i++)
```

Comment:
I'm concerned about performance, looping over every character in every key and running non-compiled regular expressions. It could be optimized using `string.IndexOfAny()`, which is vectorized. I'd like to propose something more efficient, once I better understand the escaping rules.

Reply:
This is mostly pre-existing code, I didn't attempt to reengineer it. I think all I changed here is the naming, location and addition of `__` handling, which is think is all that's needed at this point, but if you've got some ideas I'm open